ext/drreg/drreg.c

```diff
@@ -736,7 +740,12 @@ drreg_set_vector_entry(drvector_t *vec, reg_id_t reg, bool allowed)
     return DRREG_SUCCESS;
 }
 
-/* Assumes liveness info is already set up in per_thread_t */
+/* Assumes liveness info is already set up in per_thread_t. Liveness should have either
+ * been computed by a forward liveness scan upon every insertion if called outside of
+ * insertion phase, see drreg_forward_liveness_analysis(). Or if called inside insertion
+ * phase, at the end of drmgr's analysis phase once, see drreg_event_bb_analysis().
+ * We are not able to properly handle multiple users, xref i#3823.
```

Comment:
I thought drreg cannot handle multiple users if they are in the insertion phase as well as outside the phase. If I'm not mistaken, this comment makes it sound that one cannot have multiple users in general.

Reply:
But if we have multiple users just instrumenting at the insertion phase, then we should be okay right? For instance the utility function drx_insert_counter_update internally uses drreg, and a client in turn can also use drreg. Maybe mention something like "issues arise when multiple users do not all instrument inside the insert phase". Hope this helps.

Reply:
> For example, the order of clients using the insertion phase depends on their random priority they set up with drmgr. So while their order is not fixed, the drreg lazy phase is, because it's intentionally a low priority insertion phase so it runs last. 

Yea, I agree with you here. I am assuming that all user's instrumentation callbacks are executed before (with higher priority) than drreg's lazy restoration (which is way, as you pointed out, is given a low priority).

>Furthermore, the insertion phase depends on liveness computed before it. If multiple users are changing the code, how is this immune to not leading to incorrect spill/restore. Happy to chat about it.

We are assuming that all multiple users only use the instrumentation phase. In this case, none of them can make any application code changes - the app2app phase has passed, and liveliness has already been computed. Meta-instructions inserted in the instrumentation phase do not change the liveliness of registers.

Reply:
I think the second, all in insertion phase, is indeed a non-issue. I think the key is that there is no difference between a single user doing multiple insertions or multiple users, as long as the lazy phase is at the end, as long as this is serialized, and as long as they share their per-thread state, which is the case. Thanks for the thoughts!

Reply:
So I think the comment needs to be amended no?

Reply:
Just did.